File: services/security-dna/integration-libs/python/dna_middleware.py
import os
import requests
from fastapi import Request, HTTPException
from typing import Callable

TRUST_REGISTRY_URL = os.getenv("DNA_TRUST_REGISTRY_URL", "http://dna-trust-registry:8053")

async def verify_dna_middleware(request: Request, call_next: Callable):
    caller_dna = request.headers.get("X-Caller-DNA-ID")
    service_name = os.getenv("SERVICE_NAME", "python-service")
    
    if not caller_dna:
        raise HTTPException(status_code=401, detail="Missing DNA Identity / Trust Verification")
        
    # Check Trust Registry
    try:
        r = requests.get(f"{TRUST_REGISTRY_URL}/trust/verify/{caller_dna}/{service_name}", timeout=2)
        if r.status_code != 200:
            raise HTTPException(status_code=403, detail="DNA Trust Score Verification Failed")
    except requests.RequestException:
        raise HTTPException(status_code=503, detail="Trust Registry Unavailable")
        
    response = await call_next(request)
    response.headers["X-Component-DNA"] = f"{service_name}-DNA-Hash"
    return response

# Usage in FastAPI:
# app.middleware("http")(verify_dna_middleware)

File: services/security-dna/integration-libs/go/dna_middleware.go
package dnamw

import (
	"encoding/json"
	"net/http"
	"os"

	"github.com/gin-gonic/gin"
)

var trustRegistryURL = os.Getenv("DNA_TRUST_REGISTRY_URL")

// VerifyDNA is the middleware to check Trust Registry
func VerifyDNA(serviceName string) gin.HandlerFunc {
	return func(c *gin.Context) {
		callerDNA := c.GetHeader("X-Caller-DNA-ID")
		if callerDNA == "" {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Missing DNA Identity"})
			return
		}

		// Fast path cache can be added here
		resp, err := http.Get(trustRegistryURL + "/trust/verify/" + callerDNA + "/" + serviceName)
		if err != nil || resp.StatusCode != http.StatusOK {
			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{"error": "DNA Trust Score Verification Failed"})
			return
		}

		var result map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&result)
		
		c.Writer.Header().Set("X-Component-DNA", serviceName+"-DNA-Hash")
		c.Next()
	}
}
